File: code/87/Python/dp2.py
class Solution:
    def isScramble(self, s1, s2):
        if len(s1)!=len(s2): return False
        N = len(s1)
        dp = [[[False]*N for _ in range(N)] for _ in range(N+1)]
        for l in range(1, N+1):
            for i in range(N-l+1):
                for j in range(N-l+1):
                    if dp[l][i][j]: continue
                    if s1[i:i+l]==s2[j:j+l]:
                        dp[l][i][j] = True
                        continue
                    for k in range(1, l):
                        if dp[k][i][j] and dp[l-k][i+k][j+k]:
                            dp[l][i][j] = True
                            break
                        if dp[k][i][j+l-k] and dp[l-k][i+k][j]:
                            dp[l][i][j] = True
                            break
        return dp[N][0][0]

File: code/87/Java/dp2.java
class Solution {
    public boolean isScramble(String s1, String s2) {
        if (s1.length()!=s2.length()) {
            return false;
        }
        int N = s1.length();
        boolean dp[][][] = new boolean[N+1][N][N];
        boolean hasLen[] = new boolean[N+1];
        
        for (int l=1; l<N+1; l++) {
            for (int i=0; i<N-l+1; i++) {
                for (int j=0; j<N-l+1; j++) {
                    if (dp[l][i][j])
                        continue;
                    if (s1.substring(i, i+l).equals(s2.substring(j, j+l))) {
                        dp[l][i][j] = true;
                        hasLen[l] = true;
                        continue;
                    }
                    for (int k=1; k<l; k++) {
                        if (!hasLen[k] || !hasLen[l-k]) {
                            continue;
                        }
                        if (dp[k][i][j] && dp[l-k][i+k][j+k]) {
                            dp[l][i][j] = true;
                            hasLen[l] = true;
                            break;
                        }
                        if (dp[k][i][j+l-k] && dp[l-k][i+k][j]) {
                            dp[l][i][j] = true;
                            hasLen[l] = true;
                            break;
                        }
                    }
                    
                }
            }
        }
        return dp[N][0][0];
    }
}